lib/resources/aws/aws_security_group.rb

```diff
@@ -9,14 +12,170 @@ class AwsSecurityGroup < Inspec.resource(1)
+    end
+
+    # Any leftovers are unwelcome
+    unless raw_criteria.empty?
+      raise ArgumentError, "Unrecognized security group rule 'allow' criterion '#{raw_criteria.keys.first}'. Expected criteria: #{allowed_criteria.join(', ')}"
+    end
+
+    recognized_criteria
+  end
+
+  def allow__focus_on_position(rules, criteria)
+    if criteria.key?(:position)
+      idx = criteria.delete(:position)
+
+      # Normalize to a zero-based numeric index
+      case # rubocop: disable Style/EmptyCaseCondition
+      when idx.is_a?(Symbol) && idx == :first
+        idx = 0
+      when idx.is_a?(Symbol) && idx == :last
+        idx = rules.count - 1
+      when idx.is_a?(String)
+        idx = idx.to_i - 1 # We document this as 1-based, so adjust to be zero-based.
+      when idx.is_a?(Numeric)
+        idx -= 1 # We document this as 1-based, so adjust to be zero-based.
+      else
+        raise ArgumentError, "aws_security_group 'allow' 'position' criteria must be an integer or the symbols :first or :last"
+      end
+
+      unless idx < rules.count
+        raise ArgumentError, "aws_security_group 'allow' 'position' criteria #{idx+1} is out of range - there are only #{rules.count} rules for security group #{group_id}."
+      end
+
+      rules = [rules[idx]]
+    end
+    rules
+  end
+
+  def allow__match_port(rule, criteria) # rubocop: disable Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity, Metrics/AbcSize
+    if criteria[:exact] || criteria[:from_port] || criteria[:to_port]
+      # Exact match mode
+      # :port is shorthand for a single-valued port range.
+      criteria[:to_port] = criteria[:from_port] = criteria[:port] if criteria[:port]
+      to = criteria[:to_port]
+      from = criteria[:from_port]
+      # It's a match if neither criteria was specified
+      return true if to.nil? && from.nil?
+      # Normalize to integers
+      to = to.to_i unless to.nil?
+      from = from.to_i unless from.nil?
+      # It's a match if either was specified and the other was not
+      return true if rule[:to_port] == to && from.nil?
+      return true if rule[:from_port] == from && to.nil?
+      # Finally, both must match.
+      rule[:to_port] == to && rule[:from_port] == from
+    elsif !criteria[:port]
+      # port not specified, match anything
+      true
+    else
+      # Range membership mode
+      rule_from = rule[:from_port] || 0
+      rule_to = rule[:to_port] || 65535
+      (rule_from..rule_to).cover?(criteria[:port].to_i)
+    end
+  end
+
+  def allow__match_protocol(rule, criteria)
+    return true unless criteria.key?(:protocol)
+    prot = criteria[:protocol]
+    # We provide a "fluency alias" for -1 (any).
+    prot = '-1' if prot == 'any'
+
+    rule[:ip_protocol] == prot
+  end
+
+  def allow__match_ipv4_range(rule, criteria)
+    return true unless criteria.key?(:ipv4_range)
+    query = criteria[:ipv4_range]
+    query = [query] unless query.is_a?(Array)
```

Comment:
`Array(query)` should convert a non-Array to an Array and leave a variable that is already an Array alone.